cmd/gpu-kubelet-plugin/driver.go

```diff
@@ -177,6 +202,54 @@ func (d *driver) nodeUnprepareResource(ctx context.Context, claimNs kubeletplugi
 	return nil
 }
 
+func (d *driver) deviceHealthEvents(ctx context.Context, nodeName string) {
+	klog.Info("Processing device health notifications")
+	for {
+		select {
+		case <-ctx.Done():
+			klog.V(6).Info("Stop processing device health notifications")
+			return
+		case device, ok := <-d.deviceHealthMonitor.Unhealthy():
+			if !ok {
+				klog.V(6).Info("Health monitor channel closed")
+				return
+			}
+
+			klog.Warningf("Received unhealthy notification for device: %s", device.UUID())
+
+			// Mark device as unhealthy.
+			d.state.UpdateDeviceHealthStatus(device, Unhealthy)
+
+			// Republish resource slice with only healthy devices
+			// There is no remediation loop right now meaning if the unhealthy device is fixed,
+			// driver needs to be restarted to publish the resourceslice with all devices
+			var resourceSlice resourceslice.Slice
+			for _, dev := range d.state.allocatable {
+				uuid := dev.UUID()
+				if dev.IsHealthy() {
+					klog.V(6).Infof("Device: %s is healthy, added to resoureslice", uuid)
+					resourceSlice.Devices = append(resourceSlice.Devices, dev.GetDevice())
+				} else {
+					klog.Warningf("Device: %s is unhealthy, will be removed from resoureslice", uuid)
+				}
+			}
+
+			klog.V(6).Info("Rebulishing resourceslice with healthy devices")
```

Comment:
if all devices are unhealthy, we will be publishing a resource slice with empty devices. Is that intended? Also, should we synchronize this with a concurrent node prepare/unprepare calls happening (i.e prep/unprepare locks)?

Reply:
Thanks for clarifying. I suppose the advantage of holding the prep/unprep locks while updating the global device state is that any pending Prepare call will correctly fail the allocation if the device becomes unhealthy, instead of assigning an unhealthy GPU to the workload. This way, the claim remains in the pending state, allowing users to reschedule on other nodes (e.g., by re-creating the claims or through templates).